python/prima.py

```diff
@@ -0,0 +1,341 @@
+import ctypes as ct
+import ctypes.util
+import os
+from dataclasses import dataclass
+
+
+_lib_name = "primac"
+try:
+    # import from current dir
+    import numpy as np
+
+    _path = os.path.abspath(os.path.dirname(__file__))
+    _handle = np.ctypeslib.load_library(_lib_name, _path)
+except Exception:
+    # else rely on loader env variables
+    _lib_loc = ctypes.util.find_library(_lib_name)
+    if _lib_loc is None:
+        raise ValueError(f"Cannot find {_lib_name} library")
+    _handle = ct.cdll.LoadLibrary(_lib_loc)
+
+# primac callback form
+_prima_function_callback = ct.CFUNCTYPE(
+    None, ct.POINTER(ct.c_double), ct.POINTER(ct.c_double)
+)
+_prima_function_con_callback = ct.CFUNCTYPE(
+    None, ct.POINTER(ct.c_double), ct.POINTER(ct.c_double),
+    ct.POINTER(ct.c_double)
+)
+
+# prima_get_rc_string returns char*
+_handle.prima_get_rc_string.restype = ct.c_char_p
+
+# pure Python function converters
+
+
+def _prima_pure_callback(func, n):
+    def _inner(cx, cf):
+        x = [cx[i] for i in range(n)]
+        f = func(x)
+        cf[0] = f
+
+    return _prima_function_callback(_inner)
+
+
+def _prima_pure_con_callback(func, n, m):
+    def _inner(cx, cf, ccon):
+        x = [cx[i] for i in range(n)]
+        f, con = func(x)
+        cf[0] = f
+        for i in range(m):
+            ccon[i] = con[i]
+
+    return _prima_function_con_callback(_inner)
+
+
+@dataclass
+class Result:
+    """Optimization result."""
+    x: list
+    success: bool
+    status: int
+    message: str
+    fun: float
+    nfev: int
+
+
+def bobyqa(f, n, x0, xl, xu, rhobeg=1.0, rhoend=1e-3, maxfun=1000, iprint=0):
```

Comment:
Why should the user provide `n`? Isn't it possible to infer it from `len(x0)` (or `x0.size` if using `numpy.ndarray`) and then assert that it corresponds to `len(xl)` and `len(xu)`?

Other minor remarks:
1. Should `f` be renamed `fun` (in the spirit of [minimize](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html))?
2. Shouldn't the signature be `bobyqa(fun, x0, xl, xu, options)`, where `options` is a `dict` with the corresponding keys?

For these design choice, I would also need the advice of @zaikunzhang.

Reply:
Thank you Julien @jschueller  and Tom @ragonneau. I agree with what Tom said.

Reply:
yes, we could spare n, I dont know for another dict, you can already provide kwargs as a dict